### PR TITLE
Feature/unify clr object and clr value class

### PR DESCRIPTION
**Pull request by mitikov**
ClrObject and ClrValueClass has common API that highlights same logical operations are driven by these entities:

1.  Read the primitive value from known field

1.  Read the known reference field

1. Report your type

1. Report your Address

Even though similar API semantics, calling code has to deal those differently which complicates consumer code.

Extracting the common functionality allows unifying those operations and provides a background for exposing less implementation detail to calling code.

### CODE REVIEW

**Comment by mitikov** [comment]
Hi @leculver , please take a look.

Thanks & Happy Holidays!

File: src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using System;
+
+namespace Microsoft.Diagnostics.Runtime
+{
+    /// <summary>
+    /// Represents an addressable entity (class or struct) with associated type.
+    /// </summary>
+    public interface IAddressableTypedEntity: IEquatable<IAddressableTypedEntity>
+    {
+        ulong Address { get; }
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs]: Can you copy/paste the documentation strings out of ClrObject onto the properties/methods in this interface?

File: src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+        string HexAddress { get; }
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs]: Can you remove HexAddress?  It's something I shouldn't have added in the first place, and I'd like to not carry it forward.

File: src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        /// <summary>
+        /// Gets the field value from <see cref="IAddressableTypedEntity"/> with respect to field nature (either reference, or value type).
+        /// </summary>
+        /// <param name="entity">The entity to read field value from.</param>
+        /// <param name="fieldName">The name of the field to get value.</param>
+        /// <exception cref="ArgumentNullException">if entity has no type.</exception>
+        /// <exception cref="ArgumentException">if field with matching name was not found.</exception>
+        /// <returns></returns>
+        public static IAddressableTypedEntity GetFieldFrom(this IAddressableTypedEntity entity, string fieldName)
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: Please replace all instances of `var` with a concrete type in ClrMD itself.  (I don't care about the test projects.)

It makes the code a LOT more readable to show what type we are dealing with.

File: src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            var entityType = entity?.Type;
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: Nitpick:  Use ?? expression here:  `ClrType entityType = entity?.Type ?? throw new ArgumentNullException("No associated type", nameof(entity));`

Same below with GetFieldByName.

File: src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.Diagnostics.Runtime.Extensions
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: I don't think we should add a new namespace here.  No one will find this extension method.  Put this in `Microsoft.Diagnostics.Runtime` instead.

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: With that said, I do like putting in "src\Extensions", that does seem like the right place for it, even if not using a new namespace.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrValueClass.cs [hunk 6]
```diff
@@ -142,5 +142,8 @@ private ulong GetFieldAddress(string fieldName, ClrElementType element, string t
             ulong address = field.GetAddress(Address, _interior);
             return address;
         }
+
+        public bool Equals(IAddressableTypedEntity other)
+            => Address == other?.Address && Type == other?.Type;
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrValueClass.cs]: NitPick:  `=> other != null && Address == other.Address && Type == other.Type`.  This avoids creating a `ulong?` with that comparison.

File: src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs (hunk 1 above)
Comment (mitikov) [reply src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs]: I added code documentation slightly changed from ClrObject.

The difference is how exceptions are documented - I used to [follow MS guides](https://docs.microsoft.com/en-us/dotnet/csharp/codedoc) to let intelisence/resharper pick exceptions.

Please let me know if you prefer differently.

Thanks.

File: src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs (hunk 2 above)
Comment (mitikov) [reply src/Microsoft.Diagnostics.Runtime/src/Common/IAddressableTypedEntity.cs]: Yup, agree. Interface already gives Address, and HexAddress seem to be like syntax sugar to print user-friendly addresses whenever needed.

File: src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs (hunk 5 above)
Comment (mitikov) [reply src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: Yep, I had similar idea. Thanks.

File: src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs (hunk 4 above)
Comment (mitikov) [reply src/Microsoft.Diagnostics.Runtime/src/Extensions/IAddressableTypedEntityExtensions.cs]: Good one, thanks.

I try my best to keep readability as top priority, and this gotcha looks really good :)

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrValueClass.cs (hunk 6 above)
Comment (mitikov) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrValueClass.cs]: Agree, done.

**Comment by mitikov** [comment]
@leculver  Thanks for the comments, agree & resolved.

You might want to take a look at code comments and exception documentaiton style.

**Comment by leculver** [comment]
Looks good now.  I'm not back in the office until Wednesday, so I will merge the changes then.

**Pull request merged**